src/execution/window_segment_tree.cpp

```diff
@@ -1544,13 +1566,87 @@ void WindowDistinctAggregatorLocalState::Sink(DataChunk &arg_chunk, idx_t input_
 	}
 }
 
+void WindowDistinctAggregatorLocalState::ExecuteTask() {
+	auto &global_sort = *gastate.global_sort;
+	switch (stage) {
+	case PartitionSortStage::INIT:
+		//	AddLocalState is thread-safe
+		global_sort.AddLocalState(local_sort);
+		break;
+	case PartitionSortStage::MERGE: {
+		MergeSorter merge_sorter(global_sort, global_sort.buffer_manager);
+		merge_sorter.PerformInMergeRound();
+		break;
+	}
+	default:
+		break;
+	}
+
+	++gastate.tasks_completed;
+}
+
+bool WindowDistinctAggregatorGlobalState::TryPrepareNextStage(WindowDistinctAggregatorLocalState &lstate) {
+	lock_guard<mutex> stage_guard(lock);
+
+	switch (stage) {
+	case PartitionSortStage::INIT:
+		// Wait for all the local sorts to be processed
+		if (tasks_completed < locals) {
+			return false;
+		}
+		global_sort->PrepareMergePhase();
+		total_tasks = global_sort->sorted_blocks.size() / 2;
```

Comment:
Won't the parallelism slowly go down to one thread, as you only have one task if there's one sorted block?

We can merge two sorted blocks using many threads. Would it be better to take a morsel-driven approach here and parallellize based on the size of the data, i.e., number of rows?

Reply:
Because I didn't realise it did that!